attestation-service/src/verifier/snp/mod.rs

```diff
@@ -0,0 +1,272 @@
+        }
+
+        if report.vmpl != 0 {
+            return Err(anyhow!("VMPL Check Failed"));
+        }
+
+        let expected_report_data = calculate_expected_report_data(&nonce, &attestation.tee_pubkey);
+        if report.report_data != expected_report_data {
+            return Err(anyhow!("Report Data Mismatch"));
+        }
+
+        Ok(parse_tee_evidence(&report))
+    }
+}
+
+fn get_oid<const N: usize>(
+    vcek: &x509_parser::certificate::TbsCertificate,
+    oid: Oid,
+) -> Result<[u8; N]> {
+    let val = vcek
+        .get_extension_unique(&oid)?
+        .ok_or_else(|| anyhow!("Oid not found"))?
+        .value;
+
+    val.try_into().context("Oid data has wrong length.")
+}
+
+fn get_oid_int(vcek: &x509_parser::certificate::TbsCertificate, oid: Oid) -> Result<u8> {
+    let val = vcek
+        .get_extension_unique(&oid)?
+        .ok_or_else(|| anyhow!("Oid not found"))?
+        .value;
+
+    let (_, val_int) = Integer::from_der(val)?;
+    val_int.as_u8().context("Unexpected data size")
+}
+
+fn verify_report_signature(evidence: &SnpEvidence) -> Result<()> {
+    // check cert chain
+    let vcek = verify_cert_chain(&evidence.cert_chain)?;
+
+    // OpenSSL bindings do not expose custom extensions
+    // Parse the vcek using x509_parser
+    let vcek_der = &vcek.to_der()?;
+    let parsed_vcek = X509Certificate::from_der(vcek_der)?.1.tbs_certificate;
+
+    // verify vcek fields
+    // chip id
+    if get_oid::<64>(&parsed_vcek, HW_ID_OID)? != evidence.attestation_report.chip_id {
+        return Err(anyhow!("Chip ID mismatch"));
+    }
+
+    // tcb version
+    // these integer extensions are 3 bytes with the last byte as the data
+    if get_oid_int(&parsed_vcek, UCODE_SPL_OID)?
+        != evidence.attestation_report.reported_tcb.microcode
+    {
+        return Err(anyhow!("Microcode verion mismatch"));
+    }
+
+    if get_oid_int(&parsed_vcek, SNP_SPL_OID)? != evidence.attestation_report.reported_tcb.snp {
+        return Err(anyhow!("SNP verion mismatch"));
+    }
+
+    if get_oid_int(&parsed_vcek, TEE_SPL_OID)? != evidence.attestation_report.reported_tcb.tee {
+        return Err(anyhow!("TEE verion mismatch"));
+    }
+
+    if get_oid_int(&parsed_vcek, LOADER_SPL_OID)?
+        != evidence.attestation_report.reported_tcb.boot_loader
+    {
+        return Err(anyhow!("Boot loader verion mismatch"));
+    }
+
+    // verify report signature
+    let sig = ecdsa::EcdsaSig::try_from(&evidence.attestation_report.signature)?;
+    let data = &bincode::serialize(&evidence.attestation_report)?[..=0x29f];
+
+    sig.verify(data, EcKey::try_from(vcek.public_key()?)?.as_ref())
+        .context("Signature validation failed.")?;
+
+    Ok(())
+}
+
+fn load_milan_cert_chain() -> Result<(x509::X509, x509::X509)> {
+    let certs = x509::X509::stack_from_pem(include_bytes!("milan_ask_ark.pem"))?;
+
+    // ask, ark
+    Ok((certs[0].clone(), certs[1].clone()))
```

Comment:
Nit, we don't know if there are actually 2 certs in that chain.

Reply:
The cert we are reading is tracked along with the source, so we can be pretty sure of its format. Should I add a check for the length?

Reply:
> The cert we are reading is tracked along with the source, so we can be pretty sure of its format. Should I add a check for the length?

I think so, yes.